### PR TITLE
Vtcca 2565

### PR DESCRIPTION
**Pull request by jon-pike**

### CODE REVIEW

File: app/binders/GetPropertyLinksParameters.scala [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+  implicit object Binder extends QueryStringBindable[GetPropertyLinksParameters] with Cats {
+
+    override def bind(key: String, params: Map[String, Seq[String]]): Option[Either[String, GetPropertyLinksParameters]] =
+      Some(validate(params).leftMap(_.map(_.show).toList.mkString(", ")).toEither)
+
+    override def unbind(key: String, value: GetPropertyLinksParameters): String = toQueryString(value)
+
+    private def validate(params: Params): ValidationResult[GetPropertyLinksParameters] =
+      (validateString("address", params), validateString("baref", params), validateString("agent", params), validateString("status", params), validateString("sortfield", params), validateString("sortorder", params))
```
Comment (DanielConnelly) [inline app/binders/GetPropertyLinksParameters.scala]: you validate every string has max length of 1000 was this validation there before?

File: app/binders/validation/ValidationError.scala [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    case e @ AllMissingError(key, keys @ _ *) =>
+      s"""At least one of these parameters must be provided: [${(key :: keys.toList).mkString(", ")}]"""
+    case InvalidFormat(key) =>
+      s"""Invalid format for parameter "$key""""
+    case BlankQueryParameterError(key) =>
+      s"""Missing value for parameter "$key""""
+  }
+}
+
+final case class MissingError(key: String) extends ValidationError
```
Comment (DanielConnelly) [inline app/binders/validation/ValidationError.scala]: how many of these errors are actually used here

File: app/config/Global.scala [hunk 3]
```diff
@@ -36,17 +36,21 @@ import java.time.Clock
...
 import play.api.Mode.Mode
 import play.api.{Configuration, Environment}
 import uk.gov.hmrc.play.config.ServicesConfig
 
 class GuiceModule(environment: Environment,
-                  configuration: Configuration) extends AbstractModule {
+                  configuration: Configuration) extends AbstractModule with ServicesConfig {
   override def configure() = {
 
+    bind(classOf[VoaHttpClientAuth]).annotatedWith(Names.named("VoaNewAuthedBackendHttp")).to(classOf[VoaHttpClientAuth])
```
Comment (DanielConnelly) [inline app/config/Global.scala]: when the type and the definition are the same you do not require a binding here.

File: app/config/Global.scala [hunk 4]
```diff
@@ -66,5 +70,52 @@ class GuiceModule(environment: Environment,
...
+
+    bindModernisedEndpoints()
+  }
+
+  private def bindEndpoints(endpoints: Map[String, String], baseUrl: String): Unit =
+    endpoints.toList.foreach {
+      case (boundName, configPath) => bindStringWithPrefix(configPath, baseUrl, boundName)
+    }
+
+//  private def bindPropertyLinkingEndpoints(): Unit =
```
Comment (DanielConnelly) [inline app/config/Global.scala]: commented code please remove.

File: app/connectors/ExternalPropertyLinkConnector.scala [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+                                               conf: ServicesConfig) (implicit executionContext: ExecutionContext)
+  extends JsonHttpReads with OptionHttpReads with RawReads {
+
+
+
+
+  def getMyOrganisationsPropertyLinks(searchParams: GetPropertyLinksParameters, params: Option[PaginationParams])
+                                     (implicit hc: HeaderCarrier, request: ModernisedEnrichedRequest[_]): Future[Option[PropertyLinksWithAgents]] = {
+
+    println(myOrganisationsPropertyLinksUrl)
```
Comment (DanielConnelly) [inline app/connectors/ExternalPropertyLinkConnector.scala]: println .

File: app/connectors/ExternalPropertyLinkConnector.scala [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+          searchParams.agent.map("agent" -> _),
+          searchParams.status.map("status" -> _),
+          searchParams.sortfield.map("sortfield" -> _),
+          searchParams.sortorder.map("sortorder" -> _)).flatten)
+  }
+
+
+  def getMyOrganisationsPropertyLink(submissionId: String)
+                                    (implicit hc: HeaderCarrier, request: ModernisedEnrichedRequest[_]): Future[Option[OwnerPropertyLink]] =
+    http.GET[Option[OwnerPropertyLink]](myOrganisationsPropertyLinkUrl.replace("{propertyLinkId}", submissionId))
```
Comment (DanielConnelly) [inline app/connectors/ExternalPropertyLinkConnector.scala]: @PeterPerhac has moved a templating library into HMRC domain that we really should be used.

File: app/connectors/ExternalPropertyLinkConnector.scala [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    http.GET[Option[ClientPropertyLink]](myClientsPropertyLinkUrl.replace("{propertyLinkId}", submissionId))
+
+
+  def createPropertyLink(propertyLink: CreatePropertyLink)(implicit hc: HeaderCarrier, request: ModernisedEnrichedRequest[_]): Future[HttpResponse] =
+    http
+      .POST[CreatePropertyLink, HttpResponse](createPropertyLinkUrl, propertyLink, Seq())
+
+
+  private def modernisedPaginationParams(params: Option[PaginationParams]): Seq[(String, String)] =
+    params match {
```
Comment (DanielConnelly) [inline app/connectors/ExternalPropertyLinkConnector.scala]: if you are doing this sort of statement why is the field optional?

File: app/connectors/errorhandler/AttachmentClientException.scala [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package connectors.errorhandler
+
+import uk.gov.hmrc.http.HttpException
+
+class AttachmentClientException(message: String, responseCode: Int) extends HttpException(message, responseCode)
```
Comment (DanielConnelly) [inline app/connectors/errorhandler/AttachmentClientException.scala]: do we use attachments in voa-property-linking?

File: app/controllers/PropertyLinkingController.scala [hunk 9]
```diff
@@ -63,110 +66,34 @@ class PropertyLinkingController @Inject()(
...
-
-    for {
-      authResultBE <- eventualAuthResultBE
-      authResultPendingBE <- eventualAuthResultPendingBE
-      authResultApprovedNoFiltersBE <- eventualAuthResultApprovedNoFiltersBE
-    } yield Ok(Json.toJson(AgentAuthResultFE(authResultBE, authResultPendingBE.filterTotal, authResultApprovedNoFiltersBE.filterTotal)))
+  def getClientsPropertyLinks(searchParams: GetPropertyLinksParameters,
+                         paginationParams: Option[PaginationParams]) = authenticated { implicit request =>
 
+    propertyLinkService.getClientsPropertyLinks(searchParams, paginationParams) flatMap {
```
Comment (DanielConnelly) [inline app/controllers/PropertyLinkingController.scala]: does this compile?

File: app/controllers/PropertyLinkingController.scala [hunk 10]
```diff
@@ -193,30 +120,36 @@ class PropertyLinkingController @Inject()(
...
-  def clientProperty(authorisationId: Long, clientOrgId: Long, agentOrgId: Long) = authenticated { implicit request =>
-    propertyLinksConnector.get(authorisationId) flatMap {
-      case Some(authorisation) if authorisedFor(authorisation, clientOrgId, agentOrgId) => toClientProperty(authorisation) map { p => Ok(Json.toJson(p)) }
-      case _ => NotFound
-    }
+  def getMyOrganisationsAssessments(submissionId: String) = authenticated { implicit request =>
+    implicit val cache = Memoize[Long, Future[Option[GroupAccount]]]()
+
+    assessmentService.getMyOrganisationsAssessments(submissionId).fold(Ok(Json.toJson(submissionId))) {authorisation =>
+      println(Json.toJson(authorisation))
```
Comment (DanielConnelly) [inline app/controllers/PropertyLinkingController.scala]: println.

File: app/controllers/PropertyLinkingController.scala [hunk 11]
```diff
@@ -193,30 +120,36 @@ class PropertyLinkingController @Inject()(
...
+      Ok(Json.toJson(authorisation))}
   }
 
-  private def authorisedFor(authorisation: PropertiesView, clientOrgId: Long, agentOrgId: Long) = {
-    authorisation.authorisationOwnerOrganisationId == clientOrgId && authorisation.parties.exists(_.authorisedPartyOrganisationId == agentOrgId)
+  def getClientsAssessments(submissionId: String) = authenticated { implicit request =>
+    implicit val cache = Memoize[Long, Future[Option[GroupAccount]]]()
+
+    assessmentService.getClientsAssessments(submissionId).fold(Ok(Json.toJson(submissionId))) {authorisation =>
```
Comment (DanielConnelly) [inline app/controllers/PropertyLinkingController.scala]: the fold here should not throw a Ok on both.

File: app/controllers/PropertyRepresentationController.scala [hunk 12]
```diff
@@ -68,4 +71,76 @@ class PropertyRepresentationController @Inject() (
...
+        checkPermission = checkPermission,
+        challengePermission = challengePermission,
+        params = paginationParams,
+        sortfield = sortfield,
+        sortorder = sortorder,
+        address = address,
+        agent = agent).map(x => Ok(Json.toJson(x)))
+      case None =>
+        Logger.error(s"Agent details lookup failed for agentCode: $agentCode")
+        NotFound
```
Comment (DanielConnelly) [inline app/controllers/PropertyRepresentationController.scala]: Not Found like this blews up HttpVerbs sometimes, test this condition on the frontend.

File: app/controllers/PropertyRepresentationController.scala [hunk 13]
```diff
@@ -68,4 +71,76 @@ class PropertyRepresentationController @Inject() (
...
+    }
+  }
+
+
+  /** *
+    * Make two calls to the Search/Sort API
+    * the first call returns the results based on supplied filters and sortfield
+    * the second call is used only to allow us to get the count of PENDING representation requests
+    */
+  def forAgentSearchAndSort(organisationId: Long,
```
Comment (DanielConnelly) [inline app/controllers/PropertyRepresentationController.scala]: what is this method to support?

File: app/models/APIValuationHistory.scala [hunk 14]
```diff
@@ -35,6 +36,20 @@ case class APIValuationHistory(
...
+
+
 object APIValuationHistory {
   implicit val formats = Json.format[APIValuationHistory]
+
+  def apply(history: ValuationHistory) :APIValuationHistory =
+    APIValuationHistory(asstRef = history.asstRef,
+    listYear = history.listYear,
+    uarn = history.uarn,
+    effectiveDate = history.effectiveDate.get,
```
Comment (DanielConnelly) [inline app/models/APIValuationHistory.scala]: .get on option.

File: app/models/Assessment.scala [hunk 15]
```diff
@@ -29,18 +30,26 @@ case class Assessment(
                        rateableValue: Option[Long],
                        address: PropertyAddress,
                        billingAuthorityReference: String,
-                       currentFromDate: Option[LocalDate],
-                       currentToDate: Option[LocalDate],
-                       capacity: CapacityDeclaration
+                       currentFromDate: Option[LocalDate] = None,
```
Comment (DanielConnelly) [inline app/models/Assessment.scala]: why are we defaulting parameters here?

File: app/models/Assessment.scala [hunk 16]
```diff
@@ -51,9 +60,52 @@ object Assessment {
...
+    )
+  }
+
+  def fromValuationHistory(valuationHistory: ValuationHistory, authorisationId: Long) = {
+    Assessment(
+      authorisationId,
+      valuationHistory.asstRef,
+      valuationHistory.listYear,
+      valuationHistory.uarn,
+      valuationHistory.effectiveDate.getOrElse(LocalDate.now()),
```
Comment (DanielConnelly) [inline app/models/Assessment.scala]: what do we use effective date for?

File: app/models/Assessment.scala [hunk 17]
```diff
@@ -51,9 +60,52 @@ object Assessment {
...
+      valuationHistory.rateableValue.map {d => d.longValue()},
```
Comment (DanielConnelly) [inline app/models/Assessment.scala]: be careful of longValue, there is rounding involved in that.

File: app/models/modernised/CreatePropertyLink.scala [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+  implicit val format: OFormat[CreatePropertyLink] = Json.format
+
+  def apply(request: APIPropertyLinkRequest)
+    :CreatePropertyLink =
+    CreatePropertyLink(uarn = request.uarn,
+      capacity = Capacity.withName(request.authorisationOwnerCapacity),
+      startDate = request.startDate,
+      endDate = request.endDate,
+      //TODO how do we get method?
+      method = ProvidedEvidence.OTHER,
```
Comment (DanielConnelly) [inline app/models/modernised/CreatePropertyLink.scala]: does the frontend use anything with this/

File: app/models/modernised/CreatePropertyLink.scala [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+      PLsubmissionId = request.submissionId,
+      createDatetime = LocalDateTime.now(),
+      uploadedFiles = request.uploadedFiles.map(e => Evidence(formatFileName(request.submissionId, e.name), EvidenceType.withName(e.evidenceType))),
+      submissionSource = "EXTERNAL_API")
```
Comment (DanielConnelly) [inline app/models/modernised/CreatePropertyLink.scala]: not EXTERNAL_API you are DFE_UI I believe.

File: app/models/APIValuationHistory.scala (hunk 14 above)
Comment (jon-pike) [reply app/models/APIValuationHistory.scala]: Something I meant to go back and fix like the .get in the frontend. Are you OK for me to change this to option all the way through the journey and handle it in the view on the frontend?

File: app/models/Assessment.scala (hunk 16 above)
Comment (jon-pike) [reply app/models/Assessment.scala]: We use it in the DvrController.alreadySubmittedDetailedValuationRequest
If i pass an option all the way through I can handle it in the view, like I said in my previous comment.

File: app/models/Assessment.scala (hunk 17 above)
Comment (jon-pike) [reply app/models/Assessment.scala]: Are you happy for me to change rateableValue to a BigDecimal on the Frontend? Again this is used in DvrController.alreadySubmittedDetailedValuationRequest

File: app/models/modernised/CreatePropertyLink.scala (hunk 19 above)
Comment (jon-pike) [reply app/models/modernised/CreatePropertyLink.scala]: Replaced with DFE_UI

File: app/models/modernised/CreatePropertyLink.scala (hunk 18 above)
Comment (jon-pike) [reply app/models/modernised/CreatePropertyLink.scala]: Method now passed from frontend. Seems to matchup

File: app/models/Assessment.scala (hunk 17 above)
Comment (jon-pike) [reply app/models/Assessment.scala]: Left as is for the time being. Is it possible to have pence the the rateableValue?

File: app/models/Assessment.scala (hunk 16 above)
Comment (jon-pike) [reply app/models/Assessment.scala]: Passed through as option and handled in the frontend

File: app/models/Assessment.scala (hunk 15 above)
Comment (jon-pike) [reply app/models/Assessment.scala]: removed

File: app/models/APIValuationHistory.scala (hunk 14 above)
Comment (jon-pike) [reply app/models/APIValuationHistory.scala]: Pass option and handle in frontend

File: app/controllers/PropertyLinkingController.scala (hunk 10 above)
Comment (jon-pike) [reply app/controllers/PropertyLinkingController.scala]: removed

File: app/connectors/errorhandler/AttachmentClientException.scala (hunk 8 above)
Comment (jon-pike) [reply app/connectors/errorhandler/AttachmentClientException.scala]: removed

File: app/connectors/ExternalPropertyLinkConnector.scala (hunk 7 above)
Comment (jon-pike) [reply app/connectors/ExternalPropertyLinkConnector.scala]: Haven't changed this yet, but will. Not sent as optional from frontend as far as I can see

File: app/connectors/ExternalPropertyLinkConnector.scala (hunk 6 above)
Comment (jon-pike) [reply app/connectors/ExternalPropertyLinkConnector.scala]: Do you have the repo details of the library?

File: app/connectors/ExternalPropertyLinkConnector.scala (hunk 5 above)
Comment (jon-pike) [reply app/connectors/ExternalPropertyLinkConnector.scala]: removed

File: app/config/Global.scala (hunk 4 above)
Comment (jon-pike) [reply app/config/Global.scala]: removed

File: app/config/Global.scala (hunk 3 above)
Comment (jon-pike) [reply app/config/Global.scala]: removed

File: app/binders/validation/ValidationError.scala (hunk 2 above)
Comment (jon-pike) [reply app/binders/validation/ValidationError.scala]: Will look at cleaning this up.

File: app/binders/GetPropertyLinksParameters.scala (hunk 1 above)
Comment (jon-pike) [reply app/binders/GetPropertyLinksParameters.scala]: No - removed validation. Maybe still some scope for clean up.

File: app/controllers/PropertyLinkingController.scala (hunk 9 above)
Comment (jon-pike) [reply app/controllers/PropertyLinkingController.scala]: It compiles but happy to change if you see issues

File: app/controllers/PropertyLinkingController.scala (hunk 11 above)
Comment (jon-pike) [reply app/controllers/PropertyLinkingController.scala]: This was something I was going to raise, but obviously forgot. Do we want to return a NotFound in the left fold?? Had some trouble implementing this and could do with some help...

File: app/controllers/PropertyRepresentationController.scala (hunk 13 above)
Comment (jon-pike) [reply app/controllers/PropertyRepresentationController.scala]: I think we agreed to move this endpoint from PropertyLinkingController to the PropertyRepresentationController. Gets call in Dashboard YourPropertiesController.show()
and PLF RepresentationController.listRepresentationRequest()

**Pull request merged**